ddl/partition.go

```diff
@@ -155,6 +157,8 @@ func buildRangePartitionDefinitions(ctx sessionctx.Context, s *ast.CreateTableSt
 			piDef.LessThan = append(piDef.LessThan, buf.String())
 			buf.Reset()
 		}
+		// For create table statement, the created partition definition in tableInfo should be with public state.
+		piDef.State = model.StatePublic
```

Comment:
Currently even if the partition state is `delete-only`, the partition is also visible to user and the user can insert to it;

here is an example:

1. execute `alter table t add partition (partition p2 values less than (30));` in session 1;

2. in session 2:

```sql
test> show create table t;
+-------+-------------------------------------------------------------+
| Table | Create Table                                                |
+-------+-------------------------------------------------------------+
| t     | CREATE TABLE `t` (                                          |
|       |   `a` bigint(64) DEFAULT NULL,                              |
|       |   `b` bigint(64) DEFAULT NULL,                              |
|       |   KEY `idx` (`a`,`b`)                                       |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
|       | PARTITION BY RANGE ( `a` ) (                                |
|       |   PARTITION `p0` VALUES LESS THAN (3),                      |
|       |   PARTITION `p1` VALUES LESS THAN (6),                      |
|       |   PARTITION `p2` VALUES LESS THAN (30) # p2 should not be visible                     |
|       | )                                                           |
+-------+-------------------------------------------------------------+
1 row in set
Time: 0.011s
test> insert into t values (7,7); # should be faild.
Query OK, 1 row affected
Time: 0.002s
test> admin show ddl jobs;  # as you can see, the add partition job does not finish, but is is already visible to the user.
+--------+---------+------------+---------------+--------------+-----------+----------+-----------+---------------------+---------------------+---------+
| JOB_ID | DB_NAME | TABLE_NAME | JOB_TYPE      | SCHEMA_STATE | SCHEMA_ID | TABLE_ID | ROW_COUNT | START_TIME          | END_TIME            | STATE   |
+--------+---------+------------+---------------+--------------+-----------+----------+-----------+---------------------+---------------------+---------+
| 74     | test    | t          | add partition | delete only  | 1         | 67       | 0         | 2020-07-24 20:24:24 | <null>              | running |
...
```

Reply:
yes，we have noticed that, in the same issue above, Song Lingyu has done the hidden work https://github.com/pingcap/tidb/pull/18585

Reply:
BTW, instead of add`state` in `model.PartitionDefinition`, I prefer to add a `AddingDefinitions` in `model.PartitionInfo`. 

Maintain the partition state is a mental burden, if I forgot to check the partition status, It may cause a bug.

Reply:
Nice idea, some way like the column change, the related column!
How about let it be now? @imtbkcat what do you think, does check cost it too much?

Reply:
@zimulala @bb7133 How about you?

Reply:
Argument:
if we add `addingDefinitions`, that two ways:
1: let ddl add partition write directly to this field, and return. then wait for tiflash HTTP post to move it to `normalDefinition`.  shortage: DDL finish doesn't mean the partition is ready
2: let ddl add partition write to this field at the first stage, then check PD for whether the tiflash replica is ok then move it to `normalDefinition`, at last return the ddl.  shortage: this way still need two-stage(without state change) to do this, quite like what this PR looks like, just eliminating the state change and using `addingDefinition`.

Reply:
I agree with @crazycs520

Reply:
I prefer this
> 2: let ddl add partition write to this field at the first stage, then check PD for whether the tiflash replica is ok then move it to normalDefinition, at last return the ddl. shortage: this way still need two-stage(without state change) to do this, quite like what this PR looks like, just eliminating the state change and using addingDefinition.